cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/FileBatchSource.java

```diff
@@ -0,0 +1,201 @@
+import javax.annotation.Nullable;
+
+/**
+ * A {@link BatchSource} to use any distributed file system as a Source.
+ * Currently only supports S3.
+ */
+@Plugin(type = "source")
+@Name("S3")
+@Description("Batch source for S3")
+public class FileBatchSource extends BatchSource<LongWritable, Object, StructuredRecord> {
+
+  private KeyValueTable table;
+  private Date prevMinute;
+
+  private static final String REGEX_DESCRIPTION = "Regex to filter out filenames in the path. " +
+    "To use the TimeFilter, input \"timefilter\". The TimeFilter assumes that it " +
+    "are reading in files with the S3 log naming convention of YYYY-MM-DD-HH-mm-SS-Tag. The TimeFilter " +
+    "reads in files from the previous hour if the timeTable field is left blank. So if it's currently " +
+    "2015-06-16-15 (June 16th 2015, 3pm), it will read in files that contain 2015-06-16-14 in the filename. " +
+    "If the field timeTable is present, then it will read files in that haven't been read yet.";
+  private static final String ACCESS_ID_DESCRIPTION = "AccessID of the S3 source.";
+  private static final String ACCESS_KEY_DESCRIPTION = "AccessKey of the S3 source.";
+  private static final String PATH_DESCRIPTION = "Path to file(s) to be read. If a directory is specified, " +
+    "terminate the path name with a \'/\'";
+  private static final String TABLE_DESCRIPTION = "Name of the Table that keeps track of the last time files " +
+    "were read in.";
+  private static final String INPUT_FORMAT_CLASS_DESCRIPTION = "Name of the input format class. Defaults to " +
+    "textInputFormat";
+  private static final String FILE_SYSTEM_DESCRIPTION = "Distributed file system to read in from. Currently " +
+    "only supports S3.";
+
+  private static final Logger LOG = LoggerFactory.getLogger(FileBatchSource.class);
+  private S3BatchConfig config;
+
+  private static final Schema DEFAULT_SCHEMA = Schema.recordOf(
+    "event",
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("body", Schema.of(Schema.Type.STRING))
+  );
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    if (config.timeTable != null) {
+      pipelineConfigurer.createDataset(config.timeTable, KeyValueTable.class, DatasetProperties.EMPTY);
+    }
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    //calculate date one minute ago, rounded down to the nearest minute
+    prevMinute = new Date(context.getLogicalStartTime() - TimeUnit.MINUTES.toMillis(1));
+    Calendar cal = Calendar.getInstance();
+    cal.setTime(prevMinute);
+    cal.set(Calendar.SECOND, 0);
+    cal.set(Calendar.MILLISECOND, 0);
+    prevMinute = cal.getTime();
+
+    Job job = context.getHadoopJob();
+    if (!config.fileSystem.equals("S3")) {
+      LOG.warn("File system other than S3 not currently supported, using S3");
+    }
+    job.getConfiguration().set("fs.s3n.awsAccessKeyId", config.accessID);
```

Comment:
So these two lines should replace the loop/iteration I mentioned in https://github.com/caskdata/cdap/pull/2997/files#r32881761